### ISSUE TITLE
"ограничения" DisplayFactory

### ISSUE THREAD

**Issue by atherdon** [LLazyEmail/nomoretogo_email_template#116]
одним из плюсов factory design pattern в js является то что мы не создаем "новый instance" как обычно получается с классами. тоесть после создания объекта - мы не можем управлять данными. нельзя нормально сделать _get, _set и т.д.

возможно позже мы перепишем часть функционала на typescript и сможем использовать их типы данных, но сейчас есть как есть.

я изначально хотел бы добавить внутрь factory методы которые позволили бы менять settings & params. 

из важных штук которые бы хотелось иметь:
1. массив с проверками на ошибки( checksArray ), чтобы немного централизовать их. 
Например мы собираем темлпейт, при этом фейлим в футере и в хеаде. хочется чтобы при запуске сборки темлейта у нас выдавались ошибки из обоих displays и прерывалось создание темлпейта.

2. изменение сеттингов в уже готовом объекте
3. init(то что у нас сейчас делает `create`)
4. и лог метод чтобы каждый раз не писать console.log

что скажете.
я думал сделать это таким образом. создать отдельный класс в котором будут эти методы. а factory будет все собирать и возвращать уже объект со всеми settings которые мы передаем

**Comment by atherdon** [comment]
я вынес все factory в https://github.com/LLazyEmail/template-object 

пока только обсуждаем, возможно подключим Вадима и обсудим вместе next steps

**Comment by atherdon** [comment]
Да, в этом таске кстати удобно было бы обсудить улучшения текущей версии этой Factory.
если что, я готов обсуждать

**Comment by web4uartstudio** [comment]
> Да, в этом таске кстати удобно было бы обсудить улучшения текущей версии этой Factory.
> если что, я готов обсуждать

@atherdon  ввиду Ваших старых проектов и желания прийти к универсальному решению (один для всех) у меня вопрос: я тоже в данном обсуждении или на лавочке запасных пока побуду. А то как то запутался. Обычно реагировал на блоки тасков отдельно выделенные под меня.

**Comment by atherdon** [comment]
Я добавил всех в этот таск чтобы обсудить что делать дальше с factory.
Просто не всегда ж сразу есть понимание как и какую задачу делать.
Collaboration это д часть процесса.

Вадим сегодня впервые ее увидел, но при этом ему тоже нужно знать как и
зачем я ее создавал. Плюс нам нужно будет прийти к тому, чтобы Вадим вас
вел по проекту.

вс, 3 июл. 2022 г., 10:31 PM web4uartstudio ***@***.***>:

**Comment by web4uartstudio** [comment]
> Я добавил всех в этот таск чтобы обсудить что делать дальше с factory.

Я вижу пока 2 варианта:
1. Общий factory а внутри разные class каждый под свой display
2. Попробовать реализовать универсальный class с итерациями и проверками

**Comment by web4uartstudio** [comment]
Какой вариант будем реализовывать ?

**Comment by atherdon** [comment]
Пока что нас нужно закончить прошлые таски где мы выводим компоненты внутри
display.

Плюс мне сложно судить пока я не увижу вариант с кодом.

**Comment by web4uartstudio** [comment]
Ок, тогда я доделаю display по старой схеме а потом реализуем factory.

**Comment by web4uartstudio** [comment]
Каждый раз отвлекаюсь и забываю, я буду очень благодарен когда обращение ко мне будет на "ты". Я в свою очередь обещаю никому об этом не рассказывать)
@atherdon

**Comment by atherdon** [comment]
Мы сейчас уже делаем display objects with factory design pattern.

А когда мы сделаем этот таск в нас будет factory + доп фишки, которые мне
нужны

**Comment by web4uartstudio** [comment]
Не могло так получится что мы просто (без обвинений, это как версия) не добавили в главный репозиторий новую версию для которой мы делали "tableWithTwoRecipes"  ? @atherdon

**Comment by atherdon** [comment]
@web4uartstudio можете кинуть ссылки на компоненты, чтоб я не искал их? плохо пока ориентируюсь в наших папках.


в финальной версии(для использования в генераторе Markdown to Email), этот блок нам нужен в "разобранном виде".
тоесть чтобы мы могли "рендерить"/выводить внутри этого блока данные из разных рассылок.

а в каждой рассылке будет свой title, subtitle, свое описание и другие рецепты.
поидее, когда вы начнете делать display для этого блока, вы должны столкнутся с тем, что у вас есть куча данных, которые нужно засунуть в `settings` и передать в `factory` - для того чтобы этот блок мог реально выводится

**Comment by atherdon** [comment]
при этом, можно сделать вначале быстрый вариант. если у вас "контент" находится в одном файле/компоненты - мы можем сделать простой display для того, чтобы начать тестирование работы сборки целого теплейта.
но я сейчас пока еще не понял в каком состоянии унас находится - поэтому давайте вначале разберемся

**Comment by atherdon** [comment]
> Не могло так получится что мы просто (без обвинений, это как версия) не добавили в главный репозиторий новую версию для которой мы делали "tableWithTwoRecipes" ? @atherdon

я ж добавляю все ваши изменения. если какого то компонента нет - тут я скорее всего не виноват.
но т.к. в нашем репозитории есть история всех изменений - мы легко можем пройтись по ней и найти компонент который нам нужен

**Comment by atherdon** [comment]
кстати, если мы сделаем пример иерархии наших компонентов например в readme - мы потом проще сможем "бегать" по файлам, чтобы смотреть что находится внутри.
хотите такое сделать @web4uartstudio ? я могу найти шаблон и закинуть в readme.md

**Comment by web4uartstudio** [comment]
> кстати, если мы сделаем пример иерархии наших компонентов например в readme - мы потом проще сможем "бегать" по файлам, чтобы смотреть что находится внутри.
> хотите такое сделать @web4uartstudio ? я могу найти шаблон и закинуть в readme.md

Скорее всего надо будет. Но ввиду того что мы файлы гоняем по папкам появляются дубликаты а иногда не полные(в 1 из дубликатов производятся изменения), становится все сложней и сложней не запутаться.
Я принял решение полностью сейчас переделать компонент с рецептами. Приведу к одной версии, поделю на компоненты. Так будет проще. @atherdon
